eclipse-theia/theia
# Add support for `SnippetTextEdit` vscode API

#### What it does

Closes: #12022

as described [here](https://code.visualstudio.com/updates/v1_72#_improved-workspace-editing-apis) this PR implements SnippetTextEdits the same way they are implemented in vscode. This adds the api to set them on Workspaceedits and sends them as TextEdits with the insertAsSnippet flag to the monaco editor. 
Through the snippetController2 class those Edits can be applied to the current editor. 
Should  SnippetTextEdits be done on a file which editor is not currently active, they are transformed to just the text and applied as normal TextEdits. 

It sadly seems both in vscode and theia this snippetTextEdit/SnippetController2 only supports snippets with a single tabstop. So tabbing through seems to not work yet in both IDEs.

#### How to test
I built a tiny vscode extension which is downloadable from here
[snippet-text-edit-test-0.0.1.zip](https://github.com/eclipse-theia/theia/files/10361208/snippet-text-edit-test-0.0.1.zip)
or buildable from here https://github.com/jonah-iden/snippet-text-edit-test.

This just inserts a test snippetTextEdit at the start of the currently open file and a file called file2.json in the workspace root (this file needs to be created manually first).

you should see the snippet is inserted correctly at the start of the file and the cursor is moved to the tabstop and in file2.json its also inserted at the start of the file without moving the cursor.

#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
